include/nuttx/efuse/sama5_sfc_fuses.h

```diff
@@ -0,0 +1,247 @@
+/****************************************************************************
+ * include/nuttx/efuse/sama5_sfc_fuses.h
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_EFUSE_SAMA5_FUSES_H
+#define __INCLUDE_NUTTX_EFUSE_SAMA5_FUSES_H
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#if defined(ATSAMA5D2)
+#define SAM_SFC_EFUSE_MAX_LEN  544    /* Max length of sfc area. */
+#elif defined ATSAMA5D4
+#define SAM_SFC_EFUSE_MAX_LEN  512    /* Max length of sfc area. */
+#endif
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+/****************************************************************************
+ * Type Definitions
+ ****************************************************************************/
+
+#if defined(CONFIG_EFUSE) && defined(CONFIG_SAMA5_SFC)
```

Comment:
Agreed - they are in the wrong place. Will wait for other feedback and then correct this.

Reply:
> why not move the private macro/variable to source file

@xiaoxiang781216  - just to check: I assume are you referring to the `#define SAM_SFC_EFUSE_MAX_LEN` macros not the enum and/or static consts?

Reply:
No, I mean the whole stuff could be moved to arch/arm/src/sama5/. Any symbols/macros which aren't shared by multiple source file can be defined in the source file which really use them. It's always better to reduce the scope of symbols.

Reply:
I agree about the has defines.

But the static consts are the basic, default, field descriptions that can be used by an app so they need to be outside the arch directories I think?

I asked about this when I submitted the PR and @acassis noted that it ought to be in the efuse driver directory as per comments in the efuse driver files? So I left it in the sama5_sfc_efuses.h file.

If it is agreed the declarations should be moved to an arch .h file then the comments in the efuse driver files will need to be amended. Or I misunderstood the comments and they need clarifying :)

Reply:
First, I don't see any code outside the driver reference these variables before you mention that the application code will use them too. Second, these variables couple with SAMA5 tightly, I don't suggest that the application code access chip specific detail in this way since it's very hard to make the application work with other chip without the modification.

Reply:
So, it's just for test, I think you can move the header file near the driver to avoid other code use the internal code accidently at least.

Reply:
So what is meant by this, in efuse.h?

```
/* The efuse_param is used by the application to inform which field(s)
 * will be passed to the ioctl() operation.
 *   - 'efuse_desc_t *field[]': contains one or more field and it
 *     it terminated by a NULL to indicate there is no more fields.
 *     NOTE: normally the application don't need to create these fields
 *           they are mapped inside an arch efuse table and referenced
 *           in a header file inside include/nuttx/efuse/ directory.
```

Reply:
you can add chip/src to search path when app is built

Reply:
FYI @acassis said:
>  I think the efuse table currently is located at arch directory as you can see at arch/xtensa/src/esp32/esp32_efuse_table.c but there is an issue: an application is nuttx-apps/ normally cannot access files inside arch/. So I think the right place should be inside include/nuttx/efuse/

If I move that file to chip/src - or chip/src/hardware? - I will need to edit the comment that is then wrong in efuse.h then. OK?

Reply:
> Ok, look like the designer of efuse want to expose the chip specific info to application. It's strange why design in this way.

The issue to me is the writer of the app needs to know what fuses are available, but not necessarily the exact bit position(s). Also there are some fuses, relating to the boot, that are fixed in position, whereas most will be defined for the app - for example I might use 32 bits for a serial number, but someone else might only use those same bits for recording a single bit to indicate that a feature is enabled, for example.

The problem I had was there were NO examples of the fuses actually being used, so I followed the comments in efuse.h, but also the implementation for esp32c. Bound not to be 100% right!

I can see that cluttering up the include/nuttx/efuse directory with chip-related files is probably not very nice. I can see they are better located in the chip directories but there is then the issue that an app writer shouldn't "need" to refer to arch files just to understand how something works.

Perhaps the answer is:

1) Move sam_sfc_efuses.h to the chip directory as you suggest.
2) Leave the fixed fuse declarations in the file but remove the others
3) Add a comment in the sam_sfc_efuses.h file that shows an example usage of a user fuse field (for example a serial number field).
4) Edit the comment in include/nuttx/efuse/efuse.h so it doesn't say the chip fuse definitions should be there
4) Add some high level documentation to briefly explain the efuse implentation and how to make use of it in an app. There is no point me submitting my test app as an example, I don't think, as it is specific to the SAMA5D2?

I can do all the above if it makes sense - just want some advise on where such high level documentation should be put.

Thanks!